### ISSUE TITLE
Invalid SQL generated for Oracle when selecting multiple unaliased rows

### ISSUE THREAD

**Issue by alf** [jOOQ/jOOQ#13560]
Given the following query using the Oracle 18c dialect:

```
create
  .select(row(val(1)),
          row(val(1)))
  .from(values(row(1)))
  .limit(1)
  .fetch();
```

jOOQ generates the following SQL:

```
select 
  "v0" row,
  "v1" row
from (
  select 
    "x"."v0",
    "x"."v1",
    rownum "rn"
  from (
    select
      1 "v0.r1",
      1 "v1.r2"
    from (
      select 1 "c1"
      from DUAL
    ) "v"
  ) "x"
  where rownum <= (0 + 1)
)
where "rn" > 0
```

This causes Oracle to complain:

```
  "v0" row,
       *
ERROR at line 2:
ORA-00923: FROM keyword not found where expected
```

If we give our rows a name the generated SQL changes and everything works as expected:

```
create
  .select(row(val(1)).as("r1"),
          row(val(1)).as("r2"))
  .from(values(row(1)))
  .limit(1)
  .fetch();
```

```
select
  1 "r1.1",
  1 "r2.1"
from (
  select 1 "c1"
  from DUAL
) "v"
fetch next 1 rows only
```

### Versions

- jOOQ: 3.16.6
- Java: 11.0.12
- Database (include vendor): Oracle 18c XE
- OS: Linux

**Comment by lukaseder** [comment]
Thanks for your report. I've already noticed and fixed this in the context of https://github.com/jOOQ/jOOQ/issues/13069, unfortunately, not as an independent bug, so the fix wasn't backported. Will do that now.

**Comment by lukaseder** [comment]
This is the same issue as your previous report: https://github.com/jOOQ/jOOQ/issues/13095

In that previous issue, the discussion was about avoiding the now unnecessary `OFFSET .. FETCH` workaround (see https://github.com/jOOQ/jOOQ/issues/13096). But if we just don't generate the alias `row`, but something else, instead, the problem can be solved easily.

**Comment by lukaseder** [comment]
Hmm, no, the `row` alias is one problem, but it shadows another one, where `"x"."v0"` and `"x"."v1"` are invalid identifiers.

**Comment by alf** [comment]
Yes, there seems to be another issue with the generated SQL as well. :)

**Comment by alf** [comment]
I seem to bump into the issue with invalid identifiers in jOOQ 3.17.0-SNAPSHOT. I can provide a separate report for this if you wish.

**Comment by lukaseder** [comment]
> I seem to bump into the issue with invalid identifiers in jOOQ 3.17.0-SNAPSHOT. I can provide a separate report for this if you wish.

No need. I'll fix it in the context of this one here, hopefully tomorrow.

**Comment by lukaseder** [comment]
The aliasing thing is a very peculiar bug that arises from the combination of:

- Emulating `LIMIT 1` in Oracle 18c and earlier in the presence of ambiguous column names
- Using the non-multiset emulation of nested rows, which can be used in some cases

The two emulations need to communicate. In this particular case, the nested rows shouldn't be re-aliased in any different way (e.g. `"v0.r1"`) than what the `LIMIT 1` emulation expects (e.g. `"v0"`).

Given that both aliasing aim at making column names of potentially ambiguously named expressions non-ambiguous in the `LIMIT 1` emulating derived table, we should be able to just use the `"v0"` aliases and not generate any additional ones.

**Comment by lukaseder** [comment]
It's not as simple as that. It seems that the entire emulation of placing nested rows in top level selects doesn't work in this context. This test:

```java
    public void testNestedRowAmbiguousUnaliasedWithLimit() throws Exception {
        testMultisetEmulations(ctx -> {
            // [#13560] The LIMIT emulation may cause trouble in some dialects
            Record2<Record2<Integer, Integer>, Record2<Integer, Integer>> r =
            ctx.select(row(val(1), val(2)), row(val(2), val(3)))
               .limit(1)
               .fetchSingle();

            assertEquals(1, r.value1().value1());
            assertEquals(2, r.value1().value2());
            assertEquals(2, r.value2().value1());
            assertEquals(3, r.value2().value2());
        });
    }
```

Produces this wrong query:

```sql
select 
  "v0" "nested",
  "v1" "nested"
from (
  select 
    "x"."v0",
    "x"."v1",
    rownum "rn"
  from (
    select
      1 "v0.1",
      2 "v0.2",
      2 "v1.2",
      3 "v1.3"
    from DUAL
  ) "x"
  where rownum <= (0 + 1)
)
where "rn" > 0
```

The problem seems to be in the `AbstractRowAsField::forceMultisetContent` method. Whenever such nesting occurs, we should enable the `MULTISET` emulation rather than using the much older row flattening emulation. I guess that the `LIMIT 1` emulation doesn't specify the relevant subquery flags in the rendering `Context`

**Comment by lukaseder** [comment]
Yep, that's the underlying problem: https://github.com/jOOQ/jOOQ/issues/13597

**Comment by lukaseder** [comment]
#13597 seems to be quite the change for all of these `LIMIT` emulating queries. It might be worth doing eventually, if other issues arise from the lack of `Context::subquery` demarcation, but for now, I guess I could fix this particular issue also by just forcing the `DATA_MULTISET_CONTENT` flag, which is used to enforce `MULTISET` emulations inside of `MULTISET` emulations...

**Comment by lukaseder** [comment]
The problem with the `DATA_MULTISET_CONTENT` approach is that it is hard to remember whether that flag was set during SQL generation, when we fetch results from the JDBC `ResultSet`. E.g. we don't have access to the decision whether `LIMIT` was emulated.

It would obviously be possible to pass along this information, but given the many problems we've had with the flattening nested records emulation (also for embeddables), I wonder if a more thorough fix would deprecate that old emulation in favour of the JSON or XML ones from `MULTISET`: https://github.com/jOOQ/jOOQ/issues/13598

For now, I'm still trying to fix only this issue, though. https://github.com/jOOQ/jOOQ/issues/13598 has little chance of being backported to 3.16.

**Comment by lukaseder** [comment]
Anyway, I can confirm the original query works with the multiset emulation, e.g.

```sql
select 
  "v0" "nested",
  "v1" "nested"
from (
  select 
    "x"."v0",
    "x"."v1",
    rownum "rn"
  from (
    select
      json_array(nvl(
        null,
        1
      ), nvl(
        null,
        2
      ) null on null returning blob) "v0",
      json_array(nvl(
        null,
        2
      ), nvl(
        null,
        3
      ) null on null returning blob) "v1"
    from DUAL
  ) "x"
  where rownum <= (0 + 1)
)
where "rn" > 0
```

Or:

```sql
select 
  "v0" "nested",
  "v1" "nested"
from (
  select 
    "x"."v0",
    "x"."v1",
    rownum "rn"
  from (
    select
      xmlelement(
        name record,
        xmlelement(name "v0", 1),
        xmlelement(name "v1", 2)
      ) "v0",
      xmlelement(
        name record,
        xmlelement(name "v0", 2),
        xmlelement(name "v1", 3)
      ) "v1"
    from DUAL
  ) "x"
  where rownum <= (0 + 1)
)
where "rn" > 0
```

Producing:

```
+------+------+
|nested|nested|
+------+------+
|(1, 2)|(2, 3)|
+------+------+
```

**Comment by lukaseder** [comment]
I've found a way for rendering `Context` implementations to communicate with the optionally available `ExecuteContext`: https://github.com/jOOQ/jOOQ/issues/13599

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0 and 3.16.7 (https://github.com/jOOQ/jOOQ/issues/13561).
